Add go lang examples for secure coding guideline (#116)
wso2/docs-security: en/docs/security-guidelines/secure-engineering-guidelines/secure-coding-guidlines/general-recommendations-for-secure-coding.md

    ?>
    ```

### Go Specific Recommendations

Use Go's `database/sql` package with prepared statements to prevent SQL injections. The statement will be compiled and the user variables will be assigned to the query parameters at runtime. Since user variables are being set to a precompiled SQL statement, this approach avoids the possibility of SQL injections.

!!! bug error "Example Incorrect Usage"
    ```go
    customerName := r.FormValue("customerName")

>     query := "SELECT account_balance FROM user_data WHERE user_name = " + customerName

Author: Addressed with baf8fc4778251445bd3dbc6b16b6402d4b3aa657

    
    rows, err := db.Query(query)
    if err != nil {
        log.Fatal(err)
    }
    defer rows.Close()
    ```

!!! success check done "Example Correct Usage - One-Time Query"
    ```go

>     customerName := r.FormValue("customerName")

Author: Addressed with baf8fc4778251445bd3dbc6b16b6402d4b3aa657

    ```go
    customerName := r.FormValue("customerName")
    // Validations, etc...
    
    query := "SELECT account_balance FROM user_data WHERE user_name = ?"
    stmt, err := db.Prepare(query)
    if err != nil {
        return err
    }
    defer stmt.Close()
    
    rows, err := stmt.Query(customerName)
    if err != nil {
        return err
    }
    defer rows.Close()
    
    // Process rows...
    ```

When processing dynamic query segments that cannot be set as prepared statement parameters (table names, column names, ordering information, offset details), validate user input against a whitelist. This approach avoids the risk of providing the end user with the ability to append anything uncontrolled to the SQL query.

Example 1:

!!! bug error "Example Incorrect Usage"
    ```go
    customerType := r.FormValue("customerType")
    order := r.FormValue("order")
    // Validations, etc...
    
    query := "SELECT user_name, account_balance FROM user_data WHERE user_type = ? ORDER BY " + order
    rows, err := db.Query(query, customerType)
    if err != nil {
        log.Fatal(err)
    }
    defer rows.Close()
    ```

!!! success check done "Example Correct Usage"

>     ```go

Author: Addressed with baf8fc4778251445bd3dbc6b16b6402d4b3aa657

    // Package level map maintaining "user input to database column" mapping
    var validOrderColumns = map[string]string{
        "accountBalance": "account_balance",
        "userName":       "user_name",
    }
    

>     // Function definition...

Author: Addressed with baf8fc4778251445bd3dbc6b16b6402d4b3aa657

    customerType := r.FormValue("customerType")
    orderColumn := r.FormValue("orderColumn")

>     orderDirection := r.FormValue("orderDirection")

Author: Addressed with baf8fc4778251445bd3dbc6b16b6402d4b3aa657

    
    if strings.ToUpper(orderDirection) == "DESC" {
        orderDirection = "DESC"
    } else {
        orderDirection = "ASC"
    }
    
    dbColumn, exists := validOrderColumns[orderColumn]
    if !exists {
        // Handle invalid column error

>         return

Author: Addressed with baf8fc4778251445bd3dbc6b16b6402d4b3aa657

    ```

!!! success check done "Example Correct Usage"
    ```go
    import "html"
    
    fmt.Fprintf(w, `<a href="%s">Link</a>`, html.EscapeString(userInput))
    ```

Encode for JavaScript contexts

!!! bug error "Example Incorrect Usage"
    ```go
    fmt.Fprintf(w, `<script>var data = "%s";</script>`, userInput)
    ```

!!! success check done "Example Correct Usage"
    ```go
    import (
        "encoding/json"
        "html/template"
    )
    
    jsData, err := json.Marshal(userInput)
    if err != nil {
        http.Error(w, "Encoding error", http.StatusInternalServerError)
        return
    }

    tmpl := `<script>var data = {{.JSData}};</script>`
    t := template.Must(template.New("js").Parse(tmpl))
    t.Execute(w, map[string]template.JS{"JSData": template.JS(jsData)})
    ```

Encode URLs

!!! bug error "Example Incorrect Usage"
    ```go
    fmt.Fprintf(w, `<a href="%s">Click Here</a>`, userInput)
    ```

!!! success check done "Example Correct Usage"
    ```go
    escapedURL := template.URLQueryEscaper(userInput)
    fmt.Fprintf(w, `<a href="%s">Click Here</a>`, escapedURL)
    ```

#### Output Sanitization

> For Go applications, established libraries such as `bluemonday` can be used to sanitize HTML content. The `bluemonday` library allows you to define a policy for allowed HTML elements and attributes, which helps in preventing XSS attacks.

Reviewer: Let's discuss if there is actually a need for HTML sanitization in new code bases anymore. Ideally the recommendation here should be that we should use safer options like markdown if there is any such requirement, and not to use HTML inputs (since we are talking about a new line of products, we can use safer options)

### Go Specific Recommendations

When developing HTTP handlers in Go, avoid including untrusted input directly in HTTP headers. Always validate or sanitize user input before using it in headers.

!!! bug error "Example Incorrect Usage"

    ```go
    http.HandleFunc("/set-cookie", func(w http.ResponseWriter, r *http.Request) {
        author := r.URL.Query().Get("author")

        http.SetCookie(w, &http.Cookie{
            Name:  "author",
            Value: author,
            MaxAge: 3600,
        })
        w.WriteHeader(http.StatusOK)
        w.Write([]byte("Cookie set"))
    })
    ```

!!! success check done "Example Correct Usage"

    ```go
    http.HandleFunc("/set-cookie", func(w http.ResponseWriter, r *http.Request) {
        author := r.URL.Query().Get("author")

>         // Sanitize input by removing CR and LF characters
>         author = strings.ReplaceAll(author, "\r", "")
>         author = strings.ReplaceAll(author, "%0d", "")
>         author = strings.ReplaceAll(author, "\n", "")
>         author = strings.ReplaceAll(author, "%0a", "")

Reviewer: ```suggestion
        // Validate the author input. If invalid, return an error.
        if err := validateAuthorInput(author); err != nil {
            http.Error(w, "Invalid author input", http.StatusBadRequest)
            return
        }
```

Author: Addressed with 0e04e157f23d87082c1b6dbfc19b97a2e4f0ba45

        http.SetCookie(w, &http.Cookie{
            Name:  "author",
            Value: author,
            MaxAge: 3600,
        })
        w.WriteHeader(http.StatusOK)
        w.Write([]byte("Cookie set"))
    })
    ```

> !!! success check done "Example Correct Usage"
> 
>     ```go
>     http.HandleFunc("/set-cookie", func(w http.ResponseWriter, r *http.Request) {
>         author := r.URL.Query().Get("author")
>         // Reject input with CR or LF characters
>         if strings.ContainsAny(author, "\r\n") || strings.ContainsAny(author, "%0d%0a") {
>             http.Error(w, "Invalid input", http.StatusBadRequest)
>             return
>         }
> 
>         http.SetCookie(w, &http.Cookie{
>             Name:  "author",
>             Value: author,
>             MaxAge: 3600,
>         })
>         w.WriteHeader(http.StatusOK)
>         w.Write([]byte("Cookie set"))
>     })
>     ```

Reviewer: ```suggestion
```

Cookie security
:   Cookie attributes should be set properly, in order to prevent session related cookies from getting exposed over unencrypted channels.  Please refer to the documentation section [Securing Cookie](#securing-cookie). 

### Go Specific Recommendations 

> When using sessions in Go, consider setting secure and HTTP-only flags on cookies to prevent session theft via XSS attacks.

Reviewer: ```suggestion
When using sessions in Go, set secure and HTTP-only flags on cookies to prevent session theft via XSS attacks.
```

Author: Addressed with 0e04e157f23d87082c1b6dbfc19b97a2e4f0ba45

For handling sensitive information, use byte slices (`[]byte`) which are mutable, and explicitly zero out the memory before the variable goes out of scope.

!!! bug error "Example Incorrect Usage"
    ```go
    userPassword := getUserPassword(request)  // userPassword is a string, which is immutable
    ```

!!! bug error "Example Incorrect Usage"
    ```go
    userPassword := getUserPassword(request)  // Returns []byte
    valid := validateUser(username, userPassword)
    if valid {
        // Do valid actions
    }
    return valid
    ```

!!! success check done "Example Correct Usage"
    ```go
    userPassword := getUserPassword(request)  // Returns []byte
    valid := validateUser(username, userPassword)
    if valid {
        // Do valid actions
    }
    
    // Clear the password from memory
    ClearSensitiveData(userPassword)
    
    return valid
    
    // Helper function to clear sensitive data
    func ClearSensitiveData(data []byte) {
        for i := range data {
            data[i] = 0
        }
    }
    ```

> Another approach is to use a dedicated package like `crypto/subtle` which provides constant-time operations to prevent timing attacks while handling sensitive data:
> 
> !!! success check done "Example Using crypto/subtle"
>     ```go
>     import "crypto/subtle"
>     
>     // When comparing sensitive data like password hashes
>     isEqual := subtle.ConstantTimeCompare(storedHash, computedHash) == 1
>     ```
> 

Reviewer: ```suggestion
```

Not related to this section.

* Do thorough testing on the CSRF protected application to verify that there is no functional impact.

### Go Specific Recommendations

> * For Go applications, use established CSRF protection libraries such as `gorilla/csrf` or `justinas/nosurf`.

Reviewer: +1 to use `gorilla/csrf` since this seems to be the best option right now. https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html is not covering GoLang as of now too.

* Do thorough testing on the CSRF protected application to verify that there is no functional impact.

### Go Specific Recommendations

> * For Go applications, use established CSRF protection libraries such as `gorilla/csrf` or `justinas/nosurf`.

Reviewer: ```suggestion
* For Go applications, use established CSRF protection library `gorilla/csrf`. If a requirement is not met by `gorilla/csrf` library or you require custom implementation, reach security team for review.
```

Author: Addressed with 0e04e157f23d87082c1b6dbfc19b97a2e4f0ba45

        
        username := r.FormValue("username")
        password := r.FormValue("password")
        createUser(username, password)
        fmt.Fprintf(w, "User created successfully")
    }
    ```

!!! success check done "Example Recommended Usage"
    ```go
    import (
        "github.com/gorilla/csrf"
        "net/http"
    )
    
    func main() {
        CSRF := csrf.Protect(
            []byte("32-byte-long-auth-key"),
            csrf.Secure(true),
            csrf.HttpOnly(true),
        )
        
        http.HandleFunc("/create_user", createUserHandler)
        http.ListenAndServe(":8000", CSRF(http.DefaultServeMux))
    }
    
    func createUserHandler(w http.ResponseWriter, r *http.Request) {
        // CSRF validation happens automatically via the middleware
        if r.Method != http.MethodPost {
            http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
            return
        }
        
        username := r.FormValue("username")
        password := r.FormValue("password")
        
        // Process the user creation
        createUser(username, password)
        
        fmt.Fprintf(w, "User created successfully")
    }
    ```

> !!! success check done "Example Recommended Usage"
>     ```go
>     import (
>         "github.com/justinas/nosurf"
>         "net/http"
>         "html/template"
>     )
>     
>     func main() {
>         mux := http.NewServeMux()
>         mux.HandleFunc("/create_user", createUserHandler)
>         
>         // Wrap the servemux with the nosurf middleware
>         csrfHandler := nosurf.New(mux)
>         
>         // Configure cookie options
>         csrfHandler.SetBaseCookie(http.Cookie{
>             Path:     "/",
>             HttpOnly: true,
>             Secure:   true,
>             SameSite: http.SameSiteStrictMode,
>         })
>         
>         http.ListenAndServe(":8000", csrfHandler)
>     }
>     
>     func createUserHandler(w http.ResponseWriter, r *http.Request) {
>         // nosurf middleware automatically validates the token
>         if r.Method != http.MethodPost {
>             http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
>             return
>         }
>         
>         username := r.FormValue("username")
>         password := r.FormValue("password")
>         createUser(username, password)
>         fmt.Fprintf(w, "User created successfully")
>     }
>     ```

Reviewer: ```suggestion
```

Reviewer: removing since no need to give multiple options unless there is a good reason.

Author: Addressed with 0e04e157f23d87082c1b6dbfc19b97a2e4f0ba45

    }
    ```

### Go Specific Recommendations
Avoid usage of `math/rand` package for security sensitive operations and use `crypto/rand` instead.

* Always properly handle errors returned by `crypto/rand` functions as they indicate potential entropy issues.
* For generating random integers within ranges, use appropriate conversion methods that preserve randomness (avoid modulo bias).

> * Consider using established libraries like `golang.org/x/crypto` for higher-level cryptographic operations that require randomness.

Reviewer: This is part of go project: https://go.dev/wiki/X-Repositories 
+1 to use  `golang.org/x/crypto`
